hazelcast/src/main/java/com/hazelcast/core/Immutable.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.core;
+
+import com.hazelcast.spi.annotation.Beta;
+
+/**
+ * Allows notifying Hazelcast code that the object implementing this interface is effectively immutable.
+ * This may mean that it either does not have any state (e.g. pure function) or the state is
+ * not mutated at any point.
+ * This interface allows for performance optimisations where applicable such as avoiding
+ * cloning user supplied objects or cloning hazelcast internal objects supplied to the user.
+ * It is important that the user follows the rules:
+ * <ul>
+ * <li>the object must not have any state which is changed by cloning the object</li>
+ * <li>the existing state must not be changed</li>
```

Comment:
Is the first one not implied by the second one?

Reply:
The first one should cover the case where the user supplies an predicate, for example, where the result of the predicate depends on some state which is lost on cloning.
The second one should cover the case where we provide the user an internal hz object like giving him the OBJECT instance to the map store, for example, and we expect that he will play by the rules and not change it.

Reply:
I tried to make the javadoc as clear I could at the moment but I must admit the wording might not be ideal. I would definitely like to make it crystal clear that the user should "play by the rules". So I'm open to suggestions.